Tutorial/1 (#3)
juspay/todo-app: tutorial/nixify/1_nix_flake.md

# Nixify your haskell project

> This blog post series is meant to simplify Haskell development workflow for you. We'll start by discussing the challenges of working on a Haskell project without Nix and introduce [Nix flakes](https://nixos.wiki/wiki/Flakes) as the solution. If you're new to Nix, we have [resources](https://haskell.flake.page/nix-rapid) to help you get started quickly or [here](https://zero-to-nix.com) if you have time to spare. Please note that some basic knowledge of Nix expression language is assumed for this series. Throughout the series, we'll be using [todo-app](https://github.com/juspay/todo-app/tree/903c769d4bda0a8028fe3775415e9bdf29d80555), a simple Haskell project, to demonstrate how to build the project and manage runtime dependencies like [postgres](https://www.postgresql.org) and [postgREST](https://postgrest.org/en/stable) automatically without manual setup, and to highlight the benefits of Nix.

Reviewer: ```suggestion
This blog post series is meant to simplify Haskell development workflow for you. We'll start by discussing the challenges of working on a Haskell project without Nix and introduce [Nix flakes](https://nixos.wiki/wiki/Flakes) as the solution. If you're new to Nix, we have [resources](https://haskell.flake.page/nix-rapid) to help you get started quickly or [here](https://zero-to-nix.com) if you have time to spare. Please note that some basic knowledge of Nix expression language is assumed for this series. Throughout the series, we'll be using [todo-app](https://github.com/juspay/todo-app/tree/903c769d4bda0a8028fe3775415e9bdf29d80555), a simple Haskell project, to demonstrate how to build the project and manage runtime dependencies like [postgres](https://www.postgresql.org) and [postgREST](https://postgrest.org/en/stable) automatically without manual setups, and to highlight the benefits of Nix.
```

[^2]: `nixpkgs-unstable` branch is named as such because of the frequent updates it receives and doesn't imply that it is unsafe.
[^3]: `self` refers to the final state of attributes in the `outputs`. For example, `self.packages.${system}.default` refers to the attribute after assigning `pkgs.hello` to it.

Nix Flake uses two different ways to specify the location of an input flake, attribute set representation and URL-like syntax. In this post we will be using the latter for the sake of simplicity, to read more about these representations [refer here](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake.html#url-like-syntax).

For this example we will use [GNU hello](https://www.gnu.org/software/hello) package from `nixpkgs-unstable`[^2] branch of [`nixpkgs`](https://github.com/NixOS/nixpkgs) repository. 

### Outputs

- A function that accepts a required parameter `self`[^3] and optionally attributes from `inputs` as parameters. [Refer here](https://nixos.wiki/wiki/Flakes#Output_schema) for a detailed schema of `outputs`. 
- We will include `nixpkgs` as parameter to use its attributes to define our attributes in the sections below.

> - In the `let` block we also define two things, `system` ("aarch64-darwin" here) and `pkgs` consits of all the packages for `system`. In our example `system` is hardcoded for one system, [forAllSystems](https://zero-to-nix.com/concepts/flakes#system-specificity) can be used to define packages for an array of systems.

Reviewer: ```suggestion
- In the `let` block we define two values -- `system` ("aarch64-darwin" here, assuming we are on an ARM mac) and `pkgs` (referring to nixpkgs packages for `system`). 
  - In our example, as `system` is hardcoded for one system, [forAllSystems](https://zero-to-nix.com/concepts/flakes#system-specificity) can be used to define packages for an array of systems.
```

#### Packages

- Attributes in `packages.${system}` points to [derivation](https://nixos.org/manual/nix/stable/language/derivations.html) that can be used to build the package.

> - `nix build` will build the derivation of the package with the name `default`. Run `nix build .#<packageName>` to build a pacakge with a specific name.

Reviewer: ```suggestion
- `nix build` will build the derivation of the package with the name `default`. Run `nix build .#<packageName>` to build the package named "<packageName>".
```

#### Apps

> - `apps.${system}.<appName>` refers to an attrset with two required attributes `type` and `program`. The `type` attribute determines how the program should be executed, For example, "shell" for a shell script, "python" for a Python script, or "app" for an executable. On the other hand, the `program` attribute is a string that represents the path in the Nix store where the executable is located.

Reviewer: ```suggestion
- `apps.${system}.<appName>` refers to a flake app that can be run using `nix run`. It is attrset containing two keys `type` and `program`. The `type` attribute determines how the program should be executed, For example, "shell" for a shell script, "python" for a Python script, or "app" for an executable. On the other hand, the `program` attribute is a string that represents the path in the Nix store where the executable is located.
```

Reviewer: > The `type` attribute determines how the program should be executed, For example, "shell" for a shell script, "python" for a Python script, or "app" for an executable

Can you link to the docs for `type`?

Author: I couldn't find the docs for this, I got the info from the source code.

- `nix run` just takes the `executable` and based on `type` executes the `program` given by `apps.${system}.default`. Run `nix run .#<appName>` to run a specific app.

#### DevShells

- This puts you in a development shell with only the packages you need.
- You can define how your shell will look using `pkgs.mkShell`.
- `pkgs.mkShell` creates a derivation that is evaluated by the command `nix develop`
- Derivation given by `devShells.${system}.default` is evaluated by default, you can also define custom `devShell` like `devShells.${system}.mydevShell` and run it using `nix develop .#mydevShell`

> It's time to now run the following commands and see the `flake.nix` in action:

Reviewer: ```suggestion
Now you can run the following commands and see this `flake.nix` in action:
```

- `nix build`
- `nix run`
- `nix develop`

## Nixify Haskell package

We saw how to use `pkgs.hello` from `nixpkgs` in the basic example above, now it is time to define build instructions to build a package for the haskell project `todo-app`.\
Tl;dr Here's the `flake.nix` for this section:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  };
  outputs = { self, nixpkgs }:
    let
      system = "aarch64-darwin";
      pkgs = nixpkgs.legacyPackages.${system};
      overlay = final: prev: {
        todo-app = final.callCabal2nix "todo-app" ./. { };
      };
      myHaskellPackages = pkgs.haskellPackages.extend overlay;
    in
    {
      packages.${system}.default = myHaskellPackages.todo-app;
      apps.${system}.default = {
        type = "app";
        program = "${self.packages.${system}.default}/bin/todo-app";
      };
    };
}
```
Let's break it down!
### haskellPackages

> This is an attribute set that consists of all the haskell projects maintained in Nixpkgs. Since our local project will not be part of this, we have to add it. Technically you can just use  `packages.${system}.default = pkgs.${system}.haskellPackages.callCabal2nix "todo-app" ./. { };` but adding it to `haskellPackages` keeps it all in one place and consuming flakes will not have to think about where to find this package. 

Reviewer: ```suggestion
`pkgs.haskellPackages` is an attribute set consisting of all haskell packages maintained in nixpkgs. Since our local package (`todo-app`) is not already in it, we will have to add it manually. Technically you can just use  `packages.${system}.default = pkgs.${system}.haskellPackages.callCabal2nix "todo-app" ./. { };` but adding it to `haskellPackages` keeps it all in one place and any consuming flake will not have to think about where to find this package. 
```

### Overlay

> This is a function that will output an attribute consisting of the packages that you want to either add into `haskellPackages` or replace if they already exist. It takes two parameters, first parameter being the final state of `haskellPackages` after applying all the overrides and the second being the state of `haskellPackages` before applying the overrides.

Reviewer: ```suggestion
[Overlays](https://nixos.wiki/wiki/Overlays) are used to override an existing package set, such as `pkgs.haskellPackages`, and produce a new package set containing the changes we want. These changes could be either about overriding a single package in the package set (the second argument `super` references the original package set), or it could be about adding new packages to it. 
```

### callCabal2nix

> A function that takes the `name` of package (`todo-app`), `src` of cabal file (`./.`, i.e current directory), and overrides (`{ }`, i.e none) as parameters and returns a derivation that will build your package. This function internally uses ["cabal2nix"](https://github.com/NixOS/cabal2nix), a Haskell tool that generates Nix build instructions from a cabal file.

Reviewer: ```suggestion
The `callCabal2nix` function produces a Haskell package derivation given its source. This function internally uses ["cabal2nix"](https://github.com/NixOS/cabal2nix), a Haskell tool that generates Nix build instructions from a cabal file.
```

### Time to run!
After replacing the previous `flake.nix` with the current one, run the following:
- `nix build` builds the haskell project and symlinks your executable in `./result/bin`

> - `nix run` builds (if not already) and runs the `todo-app` executable

Reviewer: @shivaraj-bh 

Perhaps we should embed [asciinema](https://asciinema.org/) screencasts here for these commands? Also for the `hello` commands above?

Author: We will have a video tutorial accompanying this blog, that will resolve this!

## Nixify DevShell

> We can now build the `todo-app`! What if you want to make changes in the project and have a quick feedback on the change? [`ghcid`](https://github.com/ndmitchell/ghcid) or [`cabal repl`](https://cabal.readthedocs.io/en/3.4/cabal-commands.html#cabal-v2-repl) can be used here. These tools require a GHC loaded with packages mentioned in `build-depends` of your cabal file, which is what a `devShell` will provide (an isolated environment with custom GHC).\

Reviewer: ```suggestion
Our current flake enables us to *build* the `todo-app`. However, what if want to develop it, by adding a feature or fixing a bug? For Haskell development, we normally use [cabal](https://cabal.readthedocs.io/) and tools like [ghcid](https://github.com/ndmitchell/ghcid). These tools require a GHC loaded with packages mentioned in `build-depends` of your cabal file, which is what we want a`devShell` to provide (ie., an isolated environment with custom GHC).
```

Tl;dr Here's the `flake.nix` for this section:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  };
  outputs = { self, nixpkgs }:
    let
      system = "aarch64-darwin";
      pkgs = nixpkgs.legacyPackages.${system};
      overlay = final: prev: {
        todo-app = final.callCabal2nix "todo-app" ./. { };
      };
      myHaskellPackages = pkgs.haskellPackages.extend overlay;
    in
    {

>       devShells.${system}.default = myHaskellPackages.shellFor {
>           packages = p : [
>             p.todo-app
>           ];
>           buildInputs = with myHaskellPackages; [
>             ghcid
>             cabal-install
>             todo-app
>           ];
>         };
>     };

Reviewer: ```suggestion
      devShells.${system}.default = myHaskellPackages.shellFor {
        packages = p : [
          p.todo-app
        ];
        buildInputs = with myHaskellPackages; [
          ghcid
          cabal-install
          todo-app
        ];
      };
    };
```

}
```

### shellFor

> We will be using `shellFor`, a function defined in `haskellPackages` to set up the default shell for our project. `shellFor` is tailor made for Haskell projects, hence we choose this over `mkShell`. We only need to define two attributes, `packages` and `buildInputs`. `packages` is used to provide the Nix build instructions that we defined earlier for todo-app, so that the shell doesn't start building todo-app again. `buildInputs` is used to ensure that the specified packages are present in the `PATH` of the isolated development environment.

Reviewer: ```suggestion
In the above flake, we are using `shellFor` (a function defined in `haskellPackages` attrset) to set up the default shell for our project. `shellFor` is an abstraction over `mkShell` geared specifically for Haskell development shells. Generally, we only need to define two keys `packages` and `nativeBuildInputs`. `packages` marks which of the packages in the package set are *local* packages (to be compiled by cabal). `nativeBuildInputs` is used to ensure that the specified packages are present in the `PATH` of the isolated development environment.
```

Can you also link to the docs for these functions, like `shellFor`?

### Let's run!

- `nix develop` this will start a development shell
- `cabal repl` loads all project modules

> - `ghcid -c "cabal repl exe:todo-app"` reloads your repl session everytime there is a code change

Reviewer: Same; asciinema demos would be nice.

## Nixify external dependencies

> Most of real-world applications will rely on external dependencies like Postgres, MySQL, Redis and so on. In this section we will look at how you can start a postgres server using `nix run .#postgres` and also create the data directory local to the project.\

Reviewer: ```suggestion
So far we nixified the Haskell part of our project. But a project can also have non-Haskell dependencies, like Postgres, MySQL and redis. In this section we will specifically look at how you can start a postgres server using Nix without relying or mutating global state (outside of project directory).
```

Tl;dr Here's the `flake.nix`:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  };
  outputs = { self, nixpkgs }:
  let
    system = "aarch64-darwin";
    pkgs = nixpkgs.legacyPackages.${system};
  in
  {
    apps.${system}.postgres = {
      type = "app";
      program = 
        toString 
          (pkgs.writeShellApplication {
            name = "pg_start";
            runtimeInputs = [ pkgs.postgresql ];
            text = 
            ''
              # Initialize a database with data stored in current project dir
              [ ! -d "./data/db" ] && initdb --no-locale -D ./data/db

              postgres -D ./data/db -k "$PWD"/data
            '';

>           }) + "/bin/pg_start";

Reviewer: Maybe write it like this for clarity:

```nix
{ 
program =
  let 
    script = pkgs.writeShellApplication { ... };
  in "${script}/bin/pg_start";
}
```

    };
  };
}
```
### writeShellApplication

> - A function that generates a derivation to run the shell script in an isolated environment. 

Reviewer: Does it, in an isolated environment? The script will run any environment you run it in.

> - `runtimeInputs` are provided to the application.

Reviewer: ```suggestion
- `runtimeInputs`: packages to be made available to the shell application's PATH.
```

> - `writeShellApplication` comes with [shellcheck](https://github.com/koalaman/shellcheck) which checks the shell script for errors.

Reviewer: ```suggestion
- `writeShellApplication` uses [shellcheck](https://github.com/koalaman/shellcheck) to statically analyze your bash script for issues.
```

- `toString` function applied to `writeShellApplication` gives the path in the `nix/store` where this is located.

### Run it!

> - `nix run .#postgres` will start postgres server in your current shell.

Reviewer: Again, asciinema would be nice.

## Before vs After

- Installing postgres, starting server, loading db dump and creating database configuration for postgREST is now, `nix run .#postgres`
- Installing and running postgREST webserver is now, `nix run .#postgrest`

> - Installing cabal-install, building the project is now, `nix develop`(for development shell), `nix build` (to build the executable) and `nix run` (to run the todo-app)

Reviewer: Maybe use a Markdown table here?

# Nixify your haskell project

The purpose of this blog posts series is to simplify Haskell development workflow for you. We will begin by addressing the difficulties of working on a Haskell project without Nix, and then introduce [Nix flakes](https://nixos.wiki/wiki/Flakes) as the solution. If you're unfamiliar with Nix, we have [a quick introduction](https://haskell.flake.page/nix-rapid) available to help you get started quickly or you can take your time and explore it at [Zero to Nix](https://zero-to-nix.com). A basic understanding of the Nix expression language is assumed. Throughout the series, we will utilize a simple Haskell app called [todo-app](https://github.com/juspay/todo-app/tree/903c769d4bda0a8028fe3775415e9bdf29d80555) to illustrate how to build a Haskell project and automatically manage runtime dependencies such as databases (ie., [postgres](https://www.postgresql.org)) and and other services (here, we use [postgREST](https://postgrest.org/en/stable)), eliminating the need for any manual onboarding setup. This will allow us to highlight the advantages of using Nix.

## Why Nixify?

Why use Nix to develop a Haskell project rather than something like Stack or GHCup?

> - **Let's start with the prerequisites**: It is important to note that the Postgres server, as well as other mentioned prerequisites like postgREST, cabal-install, and GHC, do not provide the necessary information regarding their versions. Nix addresses this issue by locking these versions in `flake.lock`, ensuring that you use the same versions as the developer.

Reviewer: I feel like this should be rewritten. Looks confusing too.

Author: Yes! It's about reproducibility

Our existing flake enables us to *build* the `todo-app`. However, what if want to develop it, by adding a feature or fixing a bug? For Haskell development, we normally use [cabal](https://cabal.readthedocs.io/) and tools like [ghcid](https://github.com/ndmitchell/ghcid). These tools require a GHC environment that includes the packages specified in the `build-depends` of our cabal file. This is where `devShell` becomes useful as it provides an isolated environment with all packages required by the project, catering to our development needs.

Tl;dr Here is the `flake.nix` for this section:

```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  };
  outputs = { self, nixpkgs }:
    let
      system = "aarch64-darwin";
      pkgs = nixpkgs.legacyPackages.${system};
      overlay = final: prev: {
        todo-app = final.callCabal2nix "todo-app" ./. { };
      };
      myHaskellPackages = pkgs.haskellPackages.extend overlay;
    in
    {
      devShells.${system}.default = myHaskellPackages.shellFor {
        packages = p : [
          p.todo-app
        ];
        buildInputs = with myHaskellPackages; [
          ghcid
          cabal-install

>           todo-app

Reviewer: If you put `todo-app` in shell, running `nix develop` will recompile the project every time the .cabal or .hs file is changed.

Reviewer: I'd remove it from the post, but you don't have to re-create the asciinema.

every developer and usually takes hours or days to setup. With Nix the setup is instantaneous and reproducible,[^nix] which means any new developer
can get the development environment up and running with one command.
- **Enhanced productivity**: More time spent on writing Haskell as Nix gives a fully working development environment with `nix develop`.
- **Multi-platform**: Same configuration generally works on macOS,[^nm] Linux and WSL.

[^nix]: Considering the packages are available in Nix for the host platform.
[^nm]: Although macOS doesn't have first-class support in nixpkgs, [it is getting there](https://github.com/NixOS/nixpkgs/issues/116341).

The rest of this blog post will provide a step-by-step demonstration of how to Nixify the todo-app project.

## Introduce Flake

To begin, clone the [todo-app](https://github.com/juspay/todo-app/tree/903c769d4bda0a8028fe3775415e9bdf29d80555) repository and checkout the specified commit.

```sh
git clone https://github.com/juspay/todo-app.git
cd todo-app
git checkout 903c769d4bda0a8028fe3775415e9bdf29d80555
```

> Next, in the project's root directory, create a file named `flake.nix`. We will begin by setting up a basic template for the flake, which includes:

Reviewer 2: it will be helpful to mention adding flake.nix to git staging with `git add flake.nix`, otherwise the error `error: source tree referenced by 'git+file:///Users/hao/Sandbox/todo-app?ref=HEAD&rev=903c769d4bda0a8028fe3775415e9bdf29d80555' does not contain a '/flake.nix' file` will be confusing to newcomer

Author: It is there in the asciinema video walkthrough. Having it explicitly mentioned will be better. Thanks!
